clap_derive/src/derives/attrs.rs

```diff
@@ -194,10 +198,37 @@ impl Parser {
         let parser = Parser { kind, func };
         Sp::new(parser, parse_ident.span())
     }
+
+    pub fn validator(&self, convert_type: &Type, is_enum: Option<TokenStream>) -> TokenStream {
+        let (func, args) = if let Some(tt) = is_enum {
+            (
+                quote! {
+                    <#convert_type as ::clap::ArgEnum>::from_str
+                },
+                tt,
+            )
+        } else {
+            (self.func.clone(), quote!())
+        };
+
+        match *self.kind {
+            ParserKind::TryFromStr => quote_spanned! { func.span()=>
+                .validator(|s| {
+                    #func(s.as_str(), #args)
+                    .map(|_: #convert_type| ())
+                    .map_err(|e| e.to_string())
+                })
+            },
+            ParserKind::TryFromOsStr => quote_spanned! { func.span()=>
+                .validator_os(|s| #func(&s, #args).map(|_: #convert_type| ()))
+            },
+            _ => quote!(),
+        }
+    }
```

Comment:
I believe this should be handled by `Arg::possible_values` + `<parser>` instead of `Arg::validator` + `<parser>`

Reply:
Something for later.

Reply:
The problem is that you're trying to use validator for something that not it is supposed to be used for. This not only makes the code a little bit unclear, but also results in a degradation of the error message: possible values will not be listed. 

It doesn't changes your code much: add an extra branch here:
https://github.com/clap-rs/clap/blob/3de8af46a2f0af709d91c3af23b24054cbf74f6a/clap_derive/src/derives/into_app.rs#L199-L209

Something like:
```rust
                    ParserKind::ArgDerive=> quote! { 
                        .possible_values(<#ty as ArgMatches>::VARIANTS)
                    },
```

Reply:
And the `pub fn validator` should not be needed anymore.